src/randoop/test/ContractCheckingVisitor.java

```diff
@@ -0,0 +1,247 @@
+   * @param exceptionPredicate  the predicate to test for exceptions that are errors 
+   *
+   */
+  public ContractCheckingVisitor(List<ObjectContract> contracts, 
+                                 ExceptionPredicate exceptionPredicate) {
+    this.contracts = new ArrayList<ObjectContract>();
+    for (ObjectContract c : contracts) {
+      if (c.getArity() > 2)
+        throw new IllegalArgumentException("Visitor accepts only unary or binary contracts.");
+      this.contracts.add(c);
+    }
+    this.exceptionPredicate = exceptionPredicate;
+  }
+
+  /**
+   * {@inheritDoc}
+   * Adds checks to final statement of sequence. Looks for failure exceptions, 
+   * and violations of contracts in {@code contracts}. 
+   */
+  @Override
+  public TestChecks visit(ExecutableSequence s) {
+    ErrorRevealingChecks checks = new ErrorRevealingChecks();
+    
+    int finalIndex = s.sequence.size() - 1;
+    ExecutionOutcome finalResult = s.getResult(finalIndex);
+    
+    //If statement not executed, then something flaky
+    if (finalResult instanceof NotExecuted) {
+      throw new Error("Unexecuted final statement in sequence: " + s);
+    }
+    
+    if (finalResult instanceof ExceptionalExecution) {
+      //If there is an exception, check whether it is considered a failure
+      ExceptionalExecution exec = (ExceptionalExecution)finalResult;
+
+      if (exceptionPredicate.test(exec, s)) {
+        String exceptionName = exec.getException().getClass().getName();
+        NoExceptionCheck obs = new NoExceptionCheck(finalIndex, exceptionName);
+        checks.add(obs);
+      }
+      
+      //If exception not considered a failure, don't include checks
+      
+    } else { 
+      //Otherwise, normal execution, check contracts
+      MultiMap<Class<?>, Integer> idxmap = indicesToCheck(s);
+      for (Class<?> cls : idxmap.keySet()) {
+        for (ObjectContract c : contracts) {
+          if (c.getArity() == 1) {
+            checkUnary(s, c, idxmap.getValues(cls), checks);
+          } else {
+            checkBinary(s, c, idxmap.getValues(cls), checks);
+          }
+        }
+      }
+    }
+    return checks;
+  }
+  
+  /**
+   * Checks a binary contract over the set of values defined in the sequence,
+   * and attaches failing checks at final statement of the sequence.
+   * 
+   * @param s  the executable sequence
+   * @param c  the contract to check
+   * @param values  the set of positions defining values to check
+   * @param checks  the {@code TestChecks} to which new checks are added
+   */
+  private void checkBinary(ExecutableSequence s, ObjectContract c, Set<Integer> values, ErrorRevealingChecks checks) {
+    for (Integer i : values) {
+      for (Integer j : values) {
+
+        ExecutionOutcome result1 = s.getResult(i);
+        ExecutionOutcome result2 = s.getResult(j);
+        
+        if (! (result1 instanceof NormalExecution) 
+            && ! (result2 instanceof NormalExecution)) {
```

Comment:
Why is this "and" rather than "or"?  (I would write it as "(! ( ... and))" rather than having two negations.)